### ISSUE TITLE
Support generating self-signed leaf certificates

### ISSUE THREAD

**Issue by JamesGuthrie** [smallstep/cli#123]
I don't know if this fits into the overarching concept of smallstep, but it would be convenient for use-cases where we don't need to have a full tree of certs to be able to generate a self-signed leaf cert.

**Comment by dopey** [comment]
Hey @JamesGuthrie, thanks for posting an issue! Let me repeat the issue back to you - just to make sure I understand what feature you're asking about. 

You're able to create leaf certificates using the `step cli` but you're interested in just having self signed certs (rather than a chain).

Assuming that's right, we may have a feature that can already be useful.

```
$ bin/step certificate create roof roof.crt roof.key --profile root-ca
Please enter the password to encrypt the private key:
Your certificate has been saved in roof.crt.
Your private key has been saved in roof.key.

$ bin/step certificate inspect roof.crt
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 85018086466849770572121333623876119866 (0x3ff5e34a680710cc1fbf298220226d3a)
    Signature Algorithm: ECDSA-SHA256
        Issuer: CN=roof
        Validity
            Not Before: Aug 7 21:03:18 2019 UTC
            Not After : Aug 4 21:03:18 2029 UTC
        Subject: CN=roof
        Subject Public Key Info:
            Public Key Algorithm: ECDSA
                Public-Key: (256 bit)
                X:
                    e3:b5:23:74:ad:13:de:f5:aa:47:61:20:01:e1:64:
                    4d:43:08:19:75:d7:a4:09:0f:a8:97:6b:5c:7f:d1:
                    9a:3c
                Y:
                    7b:ce:6a:d9:09:97:33:83:5a:9c:b4:fb:7f:ab:b3:
                    83:2e:34:a3:90:74:40:12:2e:3b:6e:47:7e:8f:0a:
                    bd:0c
                Curve: P-256
        X509v3 extensions:
            X509v3 Key Usage: critical
                Certificate Sign, CRL Sign
            X509v3 Basic Constraints: critical
                CA:TRUE, pathlen:1
            X509v3 Subject Key Identifier:
                98:93:E5:F3:5A:FE:E4:15:3A:2A:B1:FE:15:C7:B4:9A:1B:0F:39:C5
            X509v3 Subject Alternative Name:
                DNS:roof

    Signature Algorithm: ECDSA-SHA256
         30:44:02:20:62:dd:b1:e9:d8:3b:b2:a8:5f:d8:46:90:21:d5:
         9f:50:05:0e:ed:52:ab:a0:e4:91:af:32:f2:58:5f:76:72:10:
         02:20:05:2c:a9:78:23:4a:0a:0b:e1:5d:1c:0c:79:58:88:8c:
         c4:dc:d9:e5:06:8a:40:16:1c:54:a8:53:30:e3:1a:1e
```

By specifying the `--profile` in the `step certificate create` command I'm able to create a self-signed `root` certificate. If you're just looking for a self signed certificate, then this might suffice.

Let me know if I've misunderstood what you're asking about. 

Also, kinda curious what you're using a self-signed `leaf` certificate for. If you don't mind elaborating on the use case, we'd love to learn about it to find out if we can improve our flows.

**Comment by JamesGuthrie** [comment]
@dopey Thanks for getting back to me. I'm aware that I can use the root profile to create a self-signed cert. The main thing that threw me off is the differences in the `X509v3 extensions` component of the certificates.

I've created two certs, one with the `root-ca` profile, and one with the `leaf` profile, signed by the "root" cert:

```
> step certificate create ss-leaf ss-leaf{.crt,.key} --profile root-ca
> step certificate create signed-leaf signed-leaf{.crt,.key} --profile leaf --ca ./ss-leaf.crt --ca-key ./ss-leaf.key
```

running `step certificate inspect` on both results in the following output:

```
> step certificate inspect ss-leaf.crt
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 230310764685716042511203065323428884195 (0xad443d64fe6ecf7605cc1b11b009aee3)
    Signature Algorithm: ECDSA-SHA256
        Issuer: CN=ss-leaf
        Validity
            Not Before: Aug 8 08:12:55 2019 UTC
            Not After : Aug 5 08:12:55 2029 UTC
        Subject: CN=ss-leaf
        Subject Public Key Info:
            Public Key Algorithm: ECDSA
                Public-Key: (256 bit)
                X:
                    78:2c:ae:ae:59:27:00:36:82:87:08:95:21:90:da:
                    eb:ba:5a:19:e8:58:98:64:83:5f:cd:21:7b:f3:1a:
                    e3:62
                Y:
                    20:3c:62:b8:57:b4:97:6e:6c:5c:8e:61:db:e1:db:
                    f0:48:2f:49:c6:81:10:f1:95:9a:89:ac:53:3c:7a:
                    9b:c1
                Curve: P-256
        X509v3 extensions:
            X509v3 Key Usage: critical
                Certificate Sign, CRL Sign
            X509v3 Basic Constraints: critical
                CA:TRUE, pathlen:1
            X509v3 Subject Key Identifier:
                C9:CA:17:64:5F:6F:31:42:49:05:71:80:40:DB:F5:B6:CF:66:C8:4D
            X509v3 Subject Alternative Name:
                DNS:ss-leaf

    Signature Algorithm: ECDSA-SHA256
         30:45:02:21:00:b3:da:10:e5:6d:3f:72:82:81:1a:4d:f3:9a:
         cc:29:6c:54:86:7b:26:7a:c0:04:e9:50:28:99:af:72:2c:76:
         f1:02:20:2f:14:bb:a7:86:4b:ba:bf:35:73:7e:b1:23:8b:58:
         8f:1d:33:02:7c:be:6f:dd:97:ed:78:a9:aa:d7:0b:59:2b
```

```
> step certificate inspect signed-leaf.crt
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 49973817674396293731414829372735948305 (0x25989b70c4807279e0a1bb2cbfff0611)
    Signature Algorithm: ECDSA-SHA256
        Issuer: CN=ss-leaf
        Validity
            Not Before: Aug 8 08:13:50 2019 UTC
            Not After : Aug 9 08:13:50 2019 UTC
        Subject: CN=signed-leaf
        Subject Public Key Info:
            Public Key Algorithm: ECDSA
                Public-Key: (256 bit)
                X:
                    9e:94:69:40:77:9a:f5:19:4a:65:20:f9:d5:71:f3:
                    4f:89:ec:6a:10:49:b9:26:70:51:e5:cf:78:ba:e3:
                    87:ed
                Y:
                    75:93:32:6a:9e:66:f2:7a:08:a3:96:6d:cd:f9:06:
                    ec:73:c5:0b:0d:88:65:5b:82:5a:af:3b:40:1e:e9:
                    91:46
                Curve: P-256
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage:
                TLS Web Server Authentication, TLS Web Client Authentication
            X509v3 Subject Key Identifier:
                64:CA:13:1B:89:1D:DD:96:0A:AF:A6:5C:1A:40:AA:2D:1E:B0:8B:3F
            X509v3 Authority Key Identifier:
                keyid:C9:CA:17:64:5F:6F:31:42:49:05:71:80:40:DB:F5:B6:CF:66:C8:4D
            X509v3 Subject Alternative Name:
                DNS:signed-leaf

    Signature Algorithm: ECDSA-SHA256
         30:45:02:20:52:54:d0:6e:d2:d4:78:63:b1:9a:cd:1c:92:8b:
         89:6e:7d:f8:a4:02:77:9b:8b:e0:f0:51:c4:db:a2:3a:86:38:
         02:21:00:d5:62:11:44:f6:5b:31:49:e3:8d:e3:84:d0:15:e6:
         df:97:9b:ff:e6:d0:b1:38:86:44:75:91:e2:23:7c:e3:81
```

Note that the `leaf` certificate has `Digital Signature, Key Encipherment` and `TLS Web Server Authentication, TLS Web Client Authentication` as key usage, while the self-signed `root-ca` certificate  has `Certificate Sign, CRL Sign`.

To be honest I'm not sure how/which clients will assert this information, but it's clear that the `root-ca` and `leaf` profiles produce different certs.

**Comment by dopey** [comment]
@JamesGuthrie you're absolutely right. The root (and intermediate) certificates are really just meant for signing and can't be used for anything else web-pki related (e.g. authenticating web traffic).

A feature request that we have received in the past is greater configurability of certificate attributes. I think this issue/request falls under the same feature. Our proposed solution is to be able to use a JSON blob/file as a template for creating a certificate. You should be able to specify the value for nearly any attribute/extension within an x.509 certificate, rather than just having our available defaults. If you have any recommendations or expectations as to how you would like this pattern to behave we'd definitely love to know. 
That feature is on our roadmap but it's not a high priority at the moment. I'll go ahead and create a new issue for that broader feature (I'll link it here), and if you could throw your support behind it (just by giving it a thumbs up or comment) that would definitely help us with planning. 

Initially, you asked if this sort of feature fits with the overarching concept of smallstep. Smallstep's goal is to create tools that do the right thing out of the box but are also highly configurable. We believe that using a certificate chain (as opposed to a single self signed cert) is the "right thing" for the vast majority of cases. We believe this to the point that we don't even support creating certificates w/out a chain (yet), because we believe that this reduces the security of the system w/out much benefit. 
Without understanding your particular use case, I can't say for certain that a certificate chain would be "better" than a self-signed leaf ... but if you have the ability to use a chain that's definitely what I would recommend.

One possible use case for a self signed leaf is that a developer is testing things locally and doesn't want to deal with a chain. If that's the case, we'd love to know, since that means we need to make our tools that deal with certificate chains even simpler (so simple that there is no benefit to using a self-signed cert). 
In that regard, we do have one feature that may be particularly useful for local development: `step certificate install root.crt` will install a certificate to your system truststore. Your web browsers and cli will implicitly trust any certificate downstream of the `root.crt` which can make local development much nicer. Your browsers won't complain about untrusted certificates and you won't have to add `--cacert` to curl (and the like) when using the command line. When you're done testing you can use `step certificate uninstall root.crt` to return the truststore to it's previous state.

tl;dr You are right! Our root and intermediate certificates cannot be used to authenticate web-pki.  A feature that makes our certificates more configurable is on our roadmap. If you can use certificate chains, we encourage you to do so (it's good practice, even if it's just for local testing).

**Comment by dopey** [comment]
Hey @JamesGuthrie, had a chance to discuss with the team. It appears that there are definitely use cases for self signed certs that I was not aware of.

We decided to add the ability to create self-signed certificates as a separate profile option (https://github.com/smallstep/cli/commit/019b712124e0a03a2b5beced6ac52206bc7dab5f):

```
$ step certificate create foo.local foo.crt foo.key --profile self-signed --subtle
```

We're requiring the `--subtle` flag because we'd still like to discourage the use of self-signed certificates unless absolutely necessary.

If you're able to build from master then you should be able to use the feature already. If you need a release (or a build for a particular operating system) let me know and we can try to cut a release candidate. 

Also, let me know if that change does/doesn't address the issue :)

**Comment by mmalone** [comment]
@JamesGuthrie we are still super interested in the specific use case here if you can share. Are you using some software or tool that _requires_ self-signed certificates or are you just thinking it'd be easier? We were speculating internally about this: whether we could persuade you to use a root certificate & signed leafs instead...

I guess if you only have two things communicating then exchanging self-signed certs is marginally simpler since you have the same amount of stuff to distribute (A's cert to B; B's cert to A for self-signed vs. Root to A; Root to B for CA) and one fewer file to manage (no root cert). But as soon as you add a third thing it seems like the distribution problem is way more annoying than the elimination of the root cert file. For self-signed with three entities you have to distribute six things (B's cert to A; C's cert to A; A's cert to B; C's cert to B; A's cert to C; B's cert to C). Woof.

Appreciate your thoughts.

**Comment by dopey** [comment]
Hey @JamesGuthrie I'm going to close this issue for the time being. Please re-open if you find the new feature to be incomplete. Cheers!

**Comment by dio** [comment]
Sorry, @dopey @mmalone this the `--subtle` flag produces the same cert as given by: https://rob-blackbourn.medium.com/how-to-use-cfssl-to-create-self-signed-certificates-d55f76ba5781? My use case is usually for e2e testing.

**Comment by dopey** [comment]
The `--subtle` flag is there to indicate that the operation can have insecure consequences. `--subtle` as in there is some subtlety to this operation, but if you know what you are doing, you should be fine.

**Comment by dio** [comment]
Thank you @dopey!
